helix-view/src/editor.rs

```diff
@@ -227,6 +229,88 @@ impl std::str::FromStr for LineNumber {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(default)]
+pub struct WhitespaceConfig {
+    pub render: WhitespaceRender,
+    pub substitutions: WhitespaceSubstitutions,
+}
+
+impl Default for WhitespaceConfig {
+    fn default() -> Self {
+        Self {
+            render: WhitespaceRender::Basic(WhitespaceRenderValue::None),
+            substitutions: WhitespaceSubstitutions::default(),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged, rename_all = "kebab-case")]
+pub enum WhitespaceRender {
+    Basic(WhitespaceRenderValue),
+    Specific {
```

Comment:
What is this Specific for? Why do we need to differentiate with basic?

Reply:
this comes from the original PR (#1218) - it's for this more advanced configuration: https://github.com/helix-editor/helix/issues/1068#issuecomment-969683249

I'm not too sure how valuable it is to have it be this configurable though. Maybe it'd be good enough to have a simple boolean that enables visible whitespace? You can also "turn off" visible whitespace for each kind of character by configuring substitutions to use spaces

```toml
[editor.whitespace.substitutions]
newline = " " # don't make newline visible
```